### PR TITLE
Add draft of the credential endpoint

### PR DESCRIPTION
**Pull request by AdamJLemmon**
Minimal draft of the credential endpoint, omitting sender constrained access token, noted as todo
Pulled the credential endpoint up to before the credential response is defined following the token endpoint response
Ref: #27

### CODE REVIEW

File: spec.md [hunk 1]
```diff
@@ -208,18 +208,85 @@ did
...
+{
+  "access_token": "eyJhbGciOiJSUzI1NiIsInR5cCI6Ikp..sHQ",
+  "token_type": "bearer",
+  "expires_in": 86400,
+  "id_token": "eyJodHRwOi8vbWF0dHIvdGVuYW50L..3Mz"
+}
+```
+# Credential Endpoint
+
+The credential endpoint is used by a Holder to request the issuance of a verifiable credential from a Credential Provder (OP).
```
Comment (tplooker) [inline spec.md:235]:
```suggestion
The credential endpoint is used by a Holder to request the issuance of a credential from a Credential Provider (OP).
```

The term credential has been defined in this spec to date but not the term `verifiable credential, im not opposed to defining it just think at this current stage `credential` is more accurate

Comment (tplooker) [reply spec.md:235]: On second thoughts maybe we should restructure this paragraph to be more inline with section [5.3 of OpenID Connect core](https://openid.net/specs/openid-connect-core-1_0.html#UserInfo)

```suggestion
The Credential Endpoint is an OAuth 2.0 Protected Resource that when called, returns Claims about the authenticated End-User in the form of a credential. To obtain the requested Claims about the End-User, the Client makes a request to the UserInfo Endpoint using an Access Token obtained through OpenID Connect Authentication whereby the the `openid_credential` scope was granted.

Communication with the Credential Endpoint MUST utilize TLS. See Section 16.17 for more information on using TLS.

The Credential Endpoint MUST support the use of HTTP POST methods defined in RFC 2616 [RFC2616].

It is recommended that the Credential Endpoint SHOULD enforce presentation of the OAuth2.0 Access Token to be sender constrained [DPOP](https://tools.ietf.org/html/draft-fett-oauth-dpop-04). However the Credential Endpoint MUST also accept Access Tokens as OAuth 2.0 Bearer Token Usage [RFC6750]. (Note: do we remove?)

The Credential Endpoint SHOULD support the use of Cross Origin Resource Sharing (CORS) [CORS] and or other methods as appropriate to enable Java Script Clients to access the endpoint.
```

File: spec.md [hunk 2]
```diff
@@ -208,18 +208,85 @@ did
...
+
+The Holder MUST use the `access_token` previously acquired in the token endpoint response from the Credential Provider as an authorization header.  It is recommended that the `access_token` be sender constrained (Note: do we require this or leave as optional based on the Credential Provider's preference, to provide a DPoP example).
```
Comment (tplooker) [inline spec.md:237]:
```suggestion
```

File: spec.md [hunk 3]
```diff
@@ -208,18 +208,85 @@ did
...
+
+## Credential Endpoint Request Parameters
+
+The Holder may provide a signed request object containing the `sub` to be used as the subject for the resulting credential.  When a `sub` claim is present within the request object an associated `sub_jwk` claim MUST also be present of which the request object MUST be signed with, therefore proving control over the `sub`.
+
+The Holder may also specify the `credential_format` they wish the credential to be returned in, given it is supported by the Credential Provider. If not specified or not supported the Credential Provider will use their preferred default format.
```
Comment (tplooker) [inline spec.md:243]:
```suggestion
The Holder may also specify the `credential_format` they wish the credential to be returned in, given it is supported by the Credential Provider. If not specified or not supported the Credential Provider will use their preferred default format. (Note if we are going to have a default we need to specify it)
```

File: spec.md [hunk 4]
```diff
@@ -208,18 +208,85 @@ did
...
+request
+: OPTIONAL. A valid OIDC signed JWT request object. The request object is used to provide a `sub` the Holder wishes to be used as the subject of the resulting credential as well as provide proof of control of that `sub`.
+
+A non-normative example of a Signed Credential request.
+
+```
+POST /credential HTTP/1.1
+Host: https://issuer.example.com
+Authorization: Bearer <access-token>
+  request=<signed-jwt-request-obj>
```
Comment (tplooker) [inline spec.md:256]: Clarifying what is the mime type of the request body here? JSON because this looks a bit like a query parameter?

Comment (AdamJLemmon) [reply spec.md:263]: Good catch, it does indeed look like a query param but was intending this to be in the body as `application/json`, update incoming :)

**Comment by AdamJLemmon** [comment]
@tplooker Anything outstanding on this one? Thanks!

File: spec.md [hunk 5]
```diff
@@ -226,18 +226,95 @@ did
...
+{
+  "access_token": "eyJhbGciOiJSUzI1NiIsInR5cCI6Ikp..sHQ",
+  "token_type": "bearer",
+  "expires_in": 86400,
+  "id_token": "eyJodHRwOi8vbWF0dHIvdGVuYW50L..3Mz"
+}
+```
+# Credential Endpoint
+
+The Credential Endpoint is an OAuth 2.0 Protected Resource that when called, returns Claims about the authenticated End-User in the form of a credential. To obtain the requested Claims about the End-User, the Client makes a request to the UserInfo Endpoint using an Access Token obtained through OpenID Connect Authentication whereby the the `openid_credential` scope was granted.
```
Comment (tplooker) [inline spec.md:253]:
```suggestion
The Credential Endpoint is an OAuth 2.0 Protected Resource that when called, returns Claims about the authenticated End-User in the form of a credential. To obtain a credential on behalf of the End-User, the Client makes a request to the Credential Endpoint using an Access Token obtained through OpenID Connect Authentication whereby the the `openid_credential` scope was granted.
```

File: spec.md [hunk 6]
```diff
@@ -226,18 +226,95 @@ did
...
+It is recommended that the Credential Endpoint SHOULD enforce presentation of the OAuth2.0 Access Token to be sender constrained [DPOP](https://tools.ietf.org/html/draft-fett-oauth-dpop-04). However the Credential Endpoint MUST also accept Access Tokens as OAuth 2.0 Bearer Token Usage [RFC6750]. (Note: do we remove?)
+
+The Credential Endpoint SHOULD support the use of Cross Origin Resource Sharing (CORS) [CORS] and or other methods as appropriate to enable Java Script Clients to access the endpoint.
+
+
+## Credential Endpoint Request Parameters
+
+The Holder may provide a signed request object containing the `sub` to be used as the subject for the resulting credential.  When a `sub` claim is present within the request object an associated `sub_jwk` claim MUST also be present of which the request object MUST be signed with, therefore proving control over the `sub`.
+
+The Holder may also specify the `credential_format` they wish the credential to be returned in, given it is supported by the Credential Provider. If not specified or not supported the Credential Provider will use their preferred default format. (Note if we are going to have a default we need to specify it)
```
Comment (tplooker) [inline spec.md:268]:
```suggestion
The Holder may also specify the `credential_format` they wish the returned credential to be formatted as. If the provider receiving the request does not support the requested credential format they it MUST return an error response, as per [xxxxx]. If the credential_format is not specified in the request SHOULD respond with their preferred or default format. (Note if we are going to have a default we need to specify it or is it at the discretion of the provider to determine this?)
```

File: spec.md [hunk 7]
```diff
@@ -226,18 +226,95 @@ did
...
+
+When a signed request is not provided the Credential Provider will use the `sub` associated with the initial `auth` request, where possible.  If a `sub` value is not available an error should be returned by the Credential Provider.
```
Comment (tplooker) [inline spec.md:270]:
```suggestion
When a signed request is not provided the Credential Provider will use the `sub` associated with the initial `auth` request, where possible.  If a `sub` value is not available the provider MUST return an error response, as per [TODO].
```

**Comment by AdamJLemmon** [comment]
Editorial changes look good to me, all committed now!

**Comment by tplooker** [comment]
Hey @AdamJLemmon looks good to merge but can you pull down the latest and re-run generating the HTML locally as the changes I have made wont be in the HTML?

**Comment by AdamJLemmon** [comment]
Thanks @tplooker  Pushed now!

**Review by tplooker — APPROVED** [review]

**Pull request merged**